src/Components/Server/src/Circuits/CircuitHost.cs

```diff
@@ -264,6 +261,36 @@ public async Task BeginInvokeDotNetFromJS(string callId, string assemblyName, st
             }
         }
 
+        public async Task OnLocationChangedAsync(string uri, bool intercepted)
+        {
+            try
+            {
+                AssertInitialized();
+                await Renderer.Dispatcher.InvokeAsync(() =>
+                {
+                    SetCurrentCircuitHost(this);
+                    Log.LocationChanging(_logger, CircuitId, uri);
+                    var navigationManager = (RemoteNavigationManager)Services.GetRequiredService<NavigationManager>();
+                    navigationManager.NotifyLocationChanged(uri, intercepted);
+                    Log.LocationChangingSucceeded(_logger, CircuitId, uri);
+                });
+            }
+            catch (Exception ex)
+            {
+                // It's up to the NavigationManager implementation to validate the URI.
+                //
+                // Note that it's also possible that setting the URI could cause a failure in code that listens
+                // to NavigationManager.LocationChanged.
+                //
+                // In either case, a well-behaved client will not send invalid URIs, and we don't really
+                // want to continue processing with the circuit if setting the URI failed inside application
+                // code. The safest thing to do is consider it a critical failure since URI is global state,
+                // and a failure means that an update to global state was partially applied.
+                Log.LocationChangingFailed(_logger, CircuitId, uri, ex);
+                UnhandledException?.Invoke(this, new UnhandledExceptionEventArgs(ex, isTerminating: false));
```

Comment:
This will allow ill-intentioned client to flood the logs with warnings.

Reply:
Javier and I discussed this offline. I'm going to revist this after https://github.com/aspnet/AspNetCore/pull/12683/files is merged - because that PR is where we're deciding the approach for this kind of thing.